File: web/src/types/forecast.ts
/**
 * Wire types for the live forecast/exhaustion state app/workers/forecast_worker.py
 * maintains. Mirrors app/schemas/forecast.py.
 */

export type ForecastMetric =
  | "cpu_percent"
  | "mem_percent"
  | "swap_percent"
  | "disk_percent"
  | "packet_loss_percent"
  | "cpu_iowait_percent"

export type ExhaustionMetric = "mem_percent" | "disk_percent"

export interface ForecastPoint {
  /** Seconds after the forecast's own computed_at. */
  offset_seconds: number
  predicted: number
  lower: number
  upper: number
}

export interface MetricForecast {
  device_id: string
  metric: ForecastMetric
  /** Which mount (disk_percent) or target (packet_loss_percent) this was
   * actually fit against — null for a single-entity metric. */
  entity: string | null
  computed_at: string
  horizon_seconds: number
  bucket_seconds: number
  /** Real history behind the fit. Null on rows written before it was recorded. */
  history_seconds: number | null
  /**
   * How far to trust this projection. A forecast now appears within minutes of
   * a device enrolling rather than after a day — the horizon is capped to the
   * history behind it, and this says so in a word.
   */
  confidence: "provisional" | "medium" | "high"
  /** Empty when there wasn't enough history to fit one — still present so
   * computed_at can say when that was last checked. */
  points: ForecastPoint[]
}

export interface ExhaustionForecast {
  device_id: string
  metric: ExhaustionMetric
  /** Which mount this was fit against for disk_percent; null for mem_percent. */
  entity: string | null
  computed_at: string
  current_value: number
  slope_per_day: number
  /** null means not projected to reach capacity on any actionable horizon —
   * either not trending upward, or too slowly to matter. */
  projected_at: string | null
}

File: backend/app/schemas/forecast.py
"""Wire schemas for the live forecast/exhaustion state app/workers/forecast_worker.py
maintains. Mirrors app/schemas/alerts.py's AnomalyBaselineOut in spirit: these
expose only *current* computed state, not history.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

Metric = Literal[
    "cpu_percent",
    "mem_percent",
    "swap_percent",
    "disk_percent",
    "packet_loss_percent",
    "cpu_iowait_percent",
]
ExhaustionMetric = Literal["mem_percent", "disk_percent"]


class ForecastPointOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    offset_seconds: int
    predicted: float
    lower: float
    upper: float


class MetricForecastOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    device_id: uuid.UUID
    metric: Metric
    #: Which mount/target this was actually fit against — null for a
    #: single-entity metric. See app/models/forecasts.py.
    entity: str | None
    computed_at: datetime
    horizon_seconds: int
    bucket_seconds: int
    #: Empty when the worker last found too little history to fit one —
    #: still returned (rather than omitted) so the caller can see when that
    #: was last checked via computed_at.
    points: list[ForecastPointOut]


class ExhaustionForecastOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    device_id: uuid.UUID
    metric: ExhaustionMetric
    #: Which mount this was fit against for disk_percent; null for mem_percent.
    entity: str | None
    computed_at: datetime
    current_value: float
    slope_per_day: float
    #: None means not projected to reach capacity on any actionable horizon
    #: — see analysis/forecast.py's ExhaustionEstimate.
    projected_at: datetime | None
